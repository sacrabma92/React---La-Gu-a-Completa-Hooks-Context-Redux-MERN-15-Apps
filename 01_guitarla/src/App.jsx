import { Header } from "./components/Header"
import { Footer } from "./components/Footer"
import { Guitar } from "./components/Guitar"
import { db } from './data/db'
import { useEffect, useState } from "react"



function App() {

  const initialCart = () => {
    const localStorageCart = localStorage.getItem('cart')
    return localStorageCart ? JSON.parse(localStorageCart) : []
  }

  const [data] = useState(db)
  const [cart, setCart] = useState(initialCart)

  const MAX_ITEMS = 5
  const MIN_ITEMS = 1

  useEffect(() => {
    localStorage.setItem('cart', JSON.stringify(cart))
  },[cart])
  

  function addToCart( item ) {
    const itemExists = cart.findIndex((guitar) => guitar.id === item.id)
    if(itemExists >= 0){
      if(cart[itemExists].quantity >= MAX_ITEMS) return
      const updatedCart = [...cart]
      updatedCart[itemExists].quantity++
      setCart(updatedCart)
    }else {
      item.quantity = 1
      setCart([...cart, item])
    }
  }

  function removeFromCart(id) {
    setCart((prevCart) => prevCart.filter(guitar => guitar.id !== id) )
  }

  function increaseQuantity(id){
    const updatedCart = cart.map( item => {
      if( item.id === id && item.quantity < MAX_ITEMS ){
        return {
          ...item,
          quantity: item.quantity + 1 
        }
      }
      return item
    })
    setCart(updatedCart)
  }

  function decreQuantity(id){
    const updatedCart = cart.map( item => {
      if( item.id === id && item.quantity > MIN_ITEMS){
        return {
          ...item,
          quantity: item.quantity - 1
        }
      }
      return item
    })
    setCart(updatedCart)
  }

  function clearCart() {
    setCart([])
  }

  return (

    <>
      <Header 
        cart={cart}
        removeFromCart={removeFromCart}
        increaseQuantity={increaseQuantity}
        decreQuantity={decreQuantity}
        clearCart={clearCart}
      />

      <main className="container-xl mt-5">
        <h2 className="text-center">Nuestra Colecci√≥n</h2>

        <div className="row mt-5">

        {data.map((guitar) => (
          <Guitar 
          key={guitar.id} 
          guitar={guitar}
          setCart={setCart}
          addToCart={addToCart}
          />
        ))}

        </div>
      </main>


      <Footer />

    </>
  )
}

export default App
