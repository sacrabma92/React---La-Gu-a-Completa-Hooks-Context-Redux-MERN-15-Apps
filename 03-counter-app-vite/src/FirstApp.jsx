
export const FirstApp = () => {
  return (
    <>
      <h1>Carlos</h1>
      <p>Soy un subtitulo</p>
    </>
  )
}