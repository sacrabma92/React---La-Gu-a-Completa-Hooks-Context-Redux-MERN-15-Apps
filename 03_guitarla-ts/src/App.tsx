import { Header } from "./components/Header"
import { Footer } from "./components/Footer"
import { Guitar } from "./components/Guitar"

import { useCart } from "./hooks/useCart"



function App() {
  // Usamos un Hook Personalizado
  const { data, cart, addToCart, removeFromCart, increaseQuantity, decreQuantity, clearCart, isEmpty, cartTotal } = useCart()

  return (

    <>
      <Header 
        cart={cart}
        removeFromCart={removeFromCart}
        increaseQuantity={increaseQuantity}
        decreQuantity={decreQuantity}
        clearCart={clearCart}
        isEmpty={isEmpty}
        cartTotal={cartTotal}
      />

      <main className="container-xl mt-5">
        <h2 className="text-center">Nuestra Colecci√≥n</h2>

        <div className="row mt-5">

        {data.map((guitar) => (
          <Guitar 
          key={guitar.id} 
          guitar={guitar}
          addToCart={addToCart}
          />
        ))}

        </div>
      </main>


      <Footer />

    </>
  )
}

export default App
