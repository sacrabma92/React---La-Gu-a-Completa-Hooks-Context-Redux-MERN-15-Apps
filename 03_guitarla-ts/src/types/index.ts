export type Guitarl = {
	id: number
	name: string
	image: string
	description: string
	price: number
}

// Heredar los types de Guitarl y a√±ade quantity
export type CartItem = Guitarl &{
  quantity: number
}

// Decemos que de Guitarl solamente tome el de tipo id
export type GuitarID = Guitarl['id']

/** Esta se usa */
// export type CartItem = Pick<Guitarl, 'id' | 'name' | 'price' > & {
//   quantity: number
// }