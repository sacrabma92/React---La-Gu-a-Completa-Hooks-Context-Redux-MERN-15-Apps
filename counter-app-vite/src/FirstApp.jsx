import PropTypes from 'prop-types'

export const FirstApp = ({ title, subtitle }) => {


  return (
    <>
      <h3> {title} </h3>
      <h1>Carlos</h1>
      <p>{subtitle}</p>
    </>
  )
}

FirstApp.propTypes = {
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string
}