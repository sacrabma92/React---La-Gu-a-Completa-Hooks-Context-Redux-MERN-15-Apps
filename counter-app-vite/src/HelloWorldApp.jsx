

export const HelloWorldApp = () => {

  return (
    <>
      <h1>Desde Rafc</h1>
    </>
  )
}


