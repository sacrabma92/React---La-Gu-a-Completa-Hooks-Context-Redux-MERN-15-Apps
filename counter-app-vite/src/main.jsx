import React from 'react'
import ReactDOM from 'react-dom/client'

import './styles.css'

import { FirstApp } from './FirstApp'
import { CounterApp } from './CounterApp'

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    {/* <HelloWorldApp /> */}
    <FirstApp title="Hola Carlos" subtitle="Soy un subtitulo" />
    {/* <CounterApp value={1992} /> */}
  </React.StrictMode>
)