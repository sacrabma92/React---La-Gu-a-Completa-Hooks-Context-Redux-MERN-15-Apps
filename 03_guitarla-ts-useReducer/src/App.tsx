import { Header } from "./components/Header"
import { Footer } from "./components/Footer"
import { Guitar } from "./components/Guitar"
import { useEffect, useReducer } from "react"

import { cartReducer, initialState } from "./reducers/cart-reducer"
import { Guitarl } from "./types"



function App() {
  
  const [state, dispatch] = useReducer(cartReducer, initialState)

  useEffect(() => {
    localStorage.setItem('cart', JSON.stringify(state.cart))
  },[state.cart])

  return (

    <>
      <Header 
        cart={state.cart}
        dispatch={dispatch}
      />

      <main className="container-xl mt-5">
        <h2 className="text-center">Nuestra Colecci√≥n</h2>

        <div className="row mt-5">

        {state.data.map((guitar : Guitarl) => (
          <Guitar 
          key={guitar.id} 
          guitar={guitar}
          dispatch={dispatch}
          />
        ))}

        </div>
      </main>


      <Footer />

    </>
  )
}

export default App
