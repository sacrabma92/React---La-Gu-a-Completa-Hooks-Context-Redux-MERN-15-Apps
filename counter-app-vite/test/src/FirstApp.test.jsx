import { render } from "@testing-library/react"
import { FirstApp } from "../../src/FirstApp"

describe('Pruebas en <FirstApp />', () => {
  test('debe hacer match con el snapshot', () => {

    const title = 'Hola, Soy Goku';
    const { container } = render(<FirstApp title={title} />)

    expect(container).toMatchSnapshot()

  })

})
