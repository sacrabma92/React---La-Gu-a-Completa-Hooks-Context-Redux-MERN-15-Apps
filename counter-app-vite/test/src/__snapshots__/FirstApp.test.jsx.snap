// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <FirstApp /> debe hacer match con el snapshot 1`] = `
<div>
  <h3>
     
    Hola, Soy Goku
     
  </h3>
  <h1>
    Carlos
  </h1>
  <p />
</div>
`;
